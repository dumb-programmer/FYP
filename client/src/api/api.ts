export function login(data: { email: string; password: string }) {
  return fetch("http://localhost:3000/login", {
    method: "POST",
    body: JSON.stringify(data),
    headers: {
      "Content-Type": "application/json",
    },
    mode: "cors",
    credentials: "include",
  });
}

export function signup(data: {
  firstName: string;
  lastName: string;
  email: string;
  password: string;
}) {
  return fetch("http://localhost:3000/signup", {
    method: "POST",
    body: JSON.stringify(data),
    headers: {
      "Content-Type": "application/json",
    },
    mode: "cors",
  });
}

export function getMessages(chatId: string, page: number) {
  return fetch(`http://localhost:3000/chats/${chatId}/messages?page=${page}`, {
    mode: "cors",
    credentials: "include",
  });
}

export function searchMessages(chatId: string, query: string, page: number) {
  return fetch(
    `http://localhost:3000/chats/${chatId}/messages?query=${query}&page=${page}`,
    {
      mode: "cors",
      credentials: "include",
    }
  );
}

export function deleteMessage(chatId: string, messageId: string) {
  return fetch(`http://localhost:3000/chats/${chatId}/messages/${messageId}`, {
    method: "DELETE",
    mode: "cors",
    credentials: "include",
  });
}

export function sendPrompt(data: { prompt: string }, chatId?: string) {
  return fetch(`http://localhost:3000/chats/${chatId}`, {
    method: "POST",
    mode: "cors",
    credentials: "include",
    body: JSON.stringify(data),
    headers: {
      "Content-Type": "application/json",
    },
  });
}

export function getChats(page: number) {
  return fetch(`http://localhost:3000/chats?page=${page}`, {
    mode: "cors",
    credentials: "include",
  });
}

export function updateChat(chatId: string, name: string) {
  return fetch(`http://localhost:3000/chats/${chatId}`, {
    method: "PATCH",
    body: JSON.stringify({ name }),
    headers: {
      "Content-Type": "application/json",
    },
    credentials: "include",
    mode: "cors",
  });
}

export function deleteChat(chatId: string) {
  return fetch(`http://localhost:3000/chats/${chatId}`, {
    method: "DELETE",
    credentials: "include",
    mode: "cors",
  });
}
