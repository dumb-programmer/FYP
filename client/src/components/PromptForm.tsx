import { sendPrompt } from "@/api/api";
import { PaperAirplaneIcon } from "@heroicons/react/16/solid";
import { useForm } from "react-hook-form";
import { useParams } from "react-router-dom";

export default function PromptForm() {
    const { handleSubmit, register, formState: { isSubmitting } } = useForm({
        defaultValues: {
            prompt: ""
        }
    });

    const { chatId } = useParams();

    return (
        <form className="flex gap-2 w-4/5" onSubmit={handleSubmit(async (data) => {
            await sendPrompt(data, chatId);
        })}>
            <input className="input input-bordered flex-1" type="text" placeholder="Prompt..." {...register("prompt")} />
            <button aria-label="send" className={`btn btn-primary ${isSubmitting ? "loading loading-spinner" : ""}`}><PaperAirplaneIcon height={20} width={20} /></button>
        </form>
    )
}
