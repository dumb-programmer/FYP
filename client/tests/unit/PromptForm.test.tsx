import React from "react";
import { render, screen } from "@testing-library/react";
import { expect, describe, it, vi } from "vitest";
import { userEvent } from "@testing-library/user-event";
import PromptForm from "../../src/components/PromptForm";
import { sendPrompt } from "../../src/api/api";
import { useParams } from "react-router-dom";

vi.mock("react-router-dom", async () => ({
    ...await vi.importActual("react-router-dom"),
    useParams: vi.fn().mockReturnValue({
        chatId: crypto.randomUUID()
    }),
}));

const mockSocket = {
    on: vi.fn(),
    off: vi.fn(),
}

vi.mock("../../src/hooks/useSocketContext", () => ({
    default: () => mockSocket
}));

vi.mock("../../src/api/api.ts", () => ({
    ...vi.importActual("../../src/api/api.ts"),
    sendPrompt: vi.fn()
}));

const renderComponent = () => render(<PromptForm />);

describe("PromptForm", () => {
    it("renders correctly", () => {
        renderComponent();

        const promptForm = screen.getByTestId(/prompt-form/i);

        expect(promptForm).toMatchSnapshot();
    });

    it("socket listeners are called on mount and cleaned up on unmount", () => {
        const { unmount } = renderComponent();

        expect(mockSocket.on).toHaveBeenCalled();
        expect(mockSocket.on).toHaveBeenCalledWith("message-start", expect.any(Function));
        expect(mockSocket.on).toHaveBeenCalled();
        expect(mockSocket.on).toHaveBeenCalledWith("message-stop", expect.any(Function));

        unmount()

        expect(mockSocket.off).toHaveBeenCalled();
        expect(mockSocket.off).toHaveBeenCalledWith("message-start", expect.any(Function));
        expect(mockSocket.off).toHaveBeenCalled();
        expect(mockSocket.off).toHaveBeenCalledWith("message-stop", expect.any(Function));
    });

    it("text input works", async () => {
        renderComponent();

        const textInput = screen.getByRole("textbox");

        await userEvent.type(textInput, "hello");

        expect(textInput.value).toEqual("hello");
    });

    it("sendPrompt is called on form submit", async () => {
        renderComponent();

        const textInput = screen.getByRole("textbox");
        const submitButton = screen.getByRole("button");

        await userEvent.type(textInput, "hello");

        await userEvent.click(submitButton);

        const { chatId } = useParams();

        expect(sendPrompt).toHaveBeenCalled();
        expect(sendPrompt).toHaveBeenCalledWith({ prompt: "hello" }, chatId);
    });
})