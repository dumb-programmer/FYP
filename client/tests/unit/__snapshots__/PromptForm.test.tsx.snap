// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`PromptForm > renders correctly 1`] = `
<form
  class="flex gap-2 w-4/5"
  data-testid="prompt-form"
>
  <input
    class="input input-bordered flex-1"
    name="prompt"
    placeholder="Prompt..."
    required=""
    type="text"
  />
  <button
    aria-label="send"
    class="btn btn-primary "
  >
    <svg
      aria-hidden="true"
      data-slot="icon"
      fill="currentColor"
      height="20"
      viewBox="0 0 16 16"
      width="20"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M2.87 2.298a.75.75 0 0 0-.812 1.021L3.39 6.624a1 1 0 0 0 .928.626H8.25a.75.75 0 0 1 0 1.5H4.318a1 1 0 0 0-.927.626l-1.333 3.305a.75.75 0 0 0 .811 1.022 24.89 24.89 0 0 0 11.668-5.115.75.75 0 0 0 0-1.175A24.89 24.89 0 0 0 2.869 2.298Z"
      />
    </svg>
  </button>
</form>
`;
